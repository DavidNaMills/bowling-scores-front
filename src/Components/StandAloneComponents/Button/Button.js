import React from 'react';
import classes from './Button.module.scss';



const Button = ({ click = () => { }, label, isDisabled = false, type = 'default', isFull=false}) => {
    let style = [classes.button];

    if(isFull){
        style = style.concat(classes.button__isFull);
    }
    if (isDisabled) {
        style = style.concat(classes.button__disabled);
    } else {

        switch (type.toLowerCase()) {
            case 'warning':
                style = style.concat(classes.button__warning);
                break;
            case 'danger':
                style = style.concat(classes.button__danger);
                break;
            default:
                style = style.concat(classes.button__default);
                break;
        }
    }

    return (
        <button onClick={click} disabled={isDisabled} className={style.join(' ')}>{label}</button>
    );
};

export default Button;