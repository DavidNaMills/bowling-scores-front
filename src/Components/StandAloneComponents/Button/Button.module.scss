@import '../../../styles/colors.scss';


.button{
    border: none;
    font-size: 16px;
    padding: 10px;
    min-width: 100px;
    font-weight: 550;

    &__disabled{
        background-color: $btn_disabled;
    }
    
    &__isFull{
        width: 100%;
    }

    &__default{
        background-color: $btn_default_bkg;
        color: $btn_font_clr_light;
        
        &:hover{
            background-color: $btn_default_bkg_active_hover;
        }
        &:active{
            background-color: $btn_default_bkg_active;
        }
    }

    &__warning{
        background-color: $btn_warning_bkg;
        color: $btn_font_clr;
        
        &:hover{
            background-color: $btn_warning_bkg_active_hover;
        }
        &:active{
            background-color: $btn_warning_bkg_active;
        }    
    }
    
    &__danger{
        background-color: $btn_danger_bkg;
        color: $btn_font_clr_light;
        
        &:hover{
            background-color: $btn_danger_bkg_active_hover;
        }
        &:active{
            background-color: $btn_danger_bkg_active;
        }    
    }

    &__lightGreen{
        background-color: $btn_lightGreen_bkg;
        color: $btn_font_clr;
        
        &:hover{
            background-color: $btn_lightGreen_bkg_active;
        }
        &:active{
            background-color: $btn_lightGreen_bkg_active_hover;
        }    
    }

    &__darkGreen{
        background-color: $btn_darkGreen_bkg;
        color: $btn_font_clr_light;
        
        &:hover{
            background-color: $btn_darkGreen_bkg_active;
        }
        &:active{
            background-color: $btn_darkGreen_bkg_active_hover;
        }    
    }

    &__blue{
        background-color: $btn_blue_bkg;
        color: $btn_font_clr_light;
        
        &:hover{
            background-color: $btn_blue_bkg_active;
        }
        &:active{
            background-color: $btn_blue_bkg_active_hover;
        }    
    }

    &__lightBlue{
        background-color: $btn_lightBlue_bkg;
        color: $btn_font_clr;
        
        &:hover{
            background-color: $btn_lightBlue_bkg_active;
        }
        &:active{
            background-color: $btn_lightBlue_bkg_active_hover;
        }    
    }

    &__darkBlue{
        background-color: $btn_darkBlue_bkg;
        color: $btn_font_clr_light;
        
        &:hover{
            background-color: $btn_darkBlue_bkg_active;
        }
        &:active{
            background-color: $btn_darkBlue_bkg_active_hover;
        }    
    }

    &__lightRed{
        background-color: $btn_lightRed_bkg;
        color: $btn_font_clr;
        
        &:hover{
            color: $btn_font_clr_light;
            background-color: $btn_lightRed_bkg_active;
        }
        &:active{
            background-color: $btn_lightRed_bkg_active_hover;
        }    
    }

    &__lightYellow{
        background-color: $btn_lightYellow_bkg;
        color: $btn_font_clr;
        
        &:hover{
            color: $btn_font_clr_light;
            background-color: $btn_lightYellow_bkg_active;
        }
        &:active{
            background-color: $btn_lightYellow_bkg_active_hover;
        }    
    }



}