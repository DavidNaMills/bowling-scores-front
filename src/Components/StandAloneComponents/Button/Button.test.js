import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Button from './Button';

//TODO: isDisabled


describe('Button component test suite', () => {
    const testLabel = 'Test label';

    describe('snapshot test suite', () => {
        test('matches default snapshot', ()=>{
            const component = renderer.create(<Button type='default' label={testLabel} />);
            const app = component.toJSON();
            expect(app).toMatchSnapshot();   
        });
        
        test('matches default snapshot with full width', ()=>{
            const component = renderer.create(<Button type='default' label={testLabel} isFull/>);
            const app = component.toJSON();
            expect(app).toMatchSnapshot();   
        });
        
        test('matches warning snapshot', ()=>{
            const component = renderer.create(<Button type='warning' label={testLabel} />);
            const app = component.toJSON();
            expect(app).toMatchSnapshot();   
        });
        
        test('matches danger snapshot', ()=>{
            const component = renderer.create(<Button type='danger' label={testLabel} />);
            const app = component.toJSON();
            expect(app).toMatchSnapshot();   
        });
    });

    describe('style test suite', () => {
        it('should add the default className', () => {
            const wrapper = shallow(<Button label={testLabel} />);
            const btn = wrapper.find('button');

            expect(btn.hasClass('button__default')).toBe(true);
            expect(btn.text()).toEqual(testLabel);
        });

        it('should add the warning className', () => {
            const wrapper = shallow(<Button label={testLabel} type='warning' />);
            const btn = wrapper.find('button');

            expect(btn.hasClass('button__warning')).toBe(true);
            expect(btn.text()).toEqual(testLabel);
        });

        it('should add the danger className', () => {
            const wrapper = shallow(<Button label={testLabel} type='danger' />);
            const btn = wrapper.find('button');

            expect(btn.hasClass('button__danger')).toBe(true);
            expect(btn.text()).toEqual(testLabel);
        });

        it('should disable the button', ()=>{
            const wrapper = shallow(<Button label={testLabel} type='danger' isDisabled/>);
            const btn = wrapper.find('button');

            expect(btn.hasClass('button__disabled')).toBe(true);
            expect(btn.props().disabled).toBeTruthy();
        })

        it('should apply full width to the button', ()=>{
            const wrapper = shallow(<Button label={testLabel} type='danger' isFull/>);
            const btn = wrapper.find('button');

            expect(btn.hasClass('button__isFull')).toBeTruthy();
        })

    });

    describe('behvioural test suite', () => {
        it('should simulate the button click', () => {
            const mockCall = jest.fn();
            const wrapper = shallow(<Button click={mockCall} label={testLabel} type='DaNgEr' />);
            const btn = wrapper.find('button');
            btn.simulate('click');

            expect(mockCall.mock.calls.length).toBe(1);
        })
    })

})