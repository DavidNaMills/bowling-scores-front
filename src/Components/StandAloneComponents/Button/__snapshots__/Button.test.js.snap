// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component test suite snapshot test suite matches danger snapshot 1`] = `
<button
  className="button button__danger"
  disabled={false}
  onClick={[Function]}
>
  Test label
</button>
`;

exports[`Button component test suite snapshot test suite matches default snapshot 1`] = `
<button
  className="button button__default"
  disabled={false}
  onClick={[Function]}
>
  Test label
</button>
`;

exports[`Button component test suite snapshot test suite matches default snapshot with full width 1`] = `
<button
  className="button button__isFull button__default"
  disabled={false}
  onClick={[Function]}
>
  Test label
</button>
`;

exports[`Button component test suite snapshot test suite matches warning snapshot 1`] = `
<button
  className="button button__warning"
  disabled={false}
  onClick={[Function]}
>
  Test label
</button>
`;
